File: python/batch_operations/src/batch_delete.py
# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
# SPDX-License-Identifier: MIT-0

"""Sequential and parallel batch DELETE for Aurora DSQL.

Deletes rows matching a condition in configurable-size batches, committing each
batch as a separate transaction to stay within the 3,000-row mutation limit.
Uses OCC retry logic with exponential backoff for serialization conflicts.

The parallel variant partitions rows across worker threads using
``abs(hashtext(id::text)::bigint) % num_workers`` so that each worker operates on a
disjoint subset, reducing OCC conflicts between workers.

NOTE: The ``table`` and ``condition`` parameters are interpolated directly into SQL.
They must come from trusted, developer-controlled sources — never from end-user
input. Use parameterized queries ($1, %s) for user-supplied values within conditions.
"""

from occ_retry import MaxRetriesExceeded, execute_with_retry

# Stop retrying a batch after this many consecutive OCC exhaustions.
MAX_CONSECUTIVE_FAILURES = 10


class IncompleteBatchError(Exception):
    """Raised when rows still match the condition after a batch operation completes."""

    def __init__(self, remaining, total_affected):
        super().__init__(
            f"{remaining} rows still match condition after processing {total_affected} rows"
        )
        self.remaining = remaining
        self.total_affected = total_affected


def batch_delete(pool, table, condition, batch_size=1000, max_retries=3, base_delay_ms=100):
    """Delete rows in batches, committing each batch as a separate transaction.

    The commit is included inside the OCC retry scope so that commit-time
    serialization failures (SQLSTATE 40001) are retried automatically.

    If a single batch exhausts its OCC retries, the loop retries that batch
    with a fresh connection (up to ``MAX_CONSECUTIVE_FAILURES`` times) instead
    of aborting the entire operation.  Previously committed batches are durable,
    so forward progress is preserved.

    Args:
        pool: A ``dsql.AuroraDSQLThreadedConnectionPool`` instance.
        table: Name of the table to delete from (trusted input only).
        condition: SQL WHERE clause without the ``WHERE`` keyword (trusted input only).
        batch_size: Number of rows to delete per transaction (default 1,000).
        max_retries: Maximum OCC retry attempts per batch (default 3).
        base_delay_ms: Base delay in milliseconds for exponential backoff (default 100).

    Returns:
        Total number of rows deleted across all batches.

    Raises:
        IncompleteBatchError: If rows still match the condition after the operation.
        MaxRetriesExceeded: If OCC retries are exhausted after MAX_CONSECUTIVE_FAILURES
            consecutive batch failures.
    """
    if batch_size < 1:
        raise ValueError("batch_size must be >= 1")

    total_deleted = 0
    consecutive_failures = 0

    while True:
        conn = pool.getconn()
        try:

            def delete_batch(conn):
                with conn.cursor() as cur:
                    cur.execute(
                        f"""DELETE FROM {table}
                        WHERE id IN (
                            SELECT id FROM {table}
                            WHERE {condition}
                            LIMIT {batch_size}
                        )"""
                    )
                    deleted = cur.rowcount
                conn.commit()
                return deleted

            deleted = execute_with_retry(
                conn, delete_batch, max_retries=max_retries, base_delay_ms=base_delay_ms
            )
            total_deleted += deleted
            consecutive_failures = 0  # reset on success
            print(f"Deleted {deleted} rows (total: {total_deleted})")

            if deleted == 0:
                break
        except MaxRetriesExceeded:
            conn.rollback()
            consecutive_failures += 1
            print(
                f"Batch OCC retries exhausted ({consecutive_failures}/"
                f"{MAX_CONSECUTIVE_FAILURES}), retrying batch with fresh connection"
            )
            if consecutive_failures >= MAX_CONSECUTIVE_FAILURES:
                raise
        except Exception:
            conn.rollback()
            raise
        finally:
            pool.putconn(conn)

    # Post-verification: ensure no matching rows remain
    conn = pool.getconn()
    try:
        with conn.cursor() as cur:
            cur.execute(f"SELECT COUNT(*) FROM {table} WHERE {condition}")
            remaining = cur.fetchone()[0]
        conn.commit()
        if remaining > 0:
            raise IncompleteBatchError(remaining, total_deleted)
    finally:
        pool.putconn(conn)

    return total_deleted

File: java/batch_operations/src/main/java/com/example/dsql/batch_operations/BatchDelete.java
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: MIT-0

package com.example.dsql.batch_operations;

import java.sql.Connection;
import java.sql.SQLException;
import java.sql.Statement;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;
import java.util.logging.Logger;
import javax.sql.DataSource;

/**
 * Sequential and parallel batch DELETE for Aurora DSQL.
 *
 * <p>NOTE: The {@code table} and {@code condition} parameters are interpolated
 * directly into SQL. They must come from trusted, developer-controlled sources —
 * never from end-user input. Use parameterized queries for user-supplied values
 * within conditions.</p>
 */
public class BatchDelete {

    private static final Logger logger = Logger.getLogger(BatchDelete.class.getName());

    /** Stop retrying a batch after this many consecutive OCC exhaustions. */
    private static final int MAX_CONSECUTIVE_FAILURES = 10;

    /**
     * Thrown when rows still match the condition after a batch operation completes.
     */
    public static class IncompleteBatchException extends SQLException {
        private final int remaining;
        private final int totalAffected;

        public IncompleteBatchException(int remaining, int totalAffected) {
            super(remaining + " rows still match condition after deleting " + totalAffected + " rows");
            this.remaining = remaining;
            this.totalAffected = totalAffected;
        }

        public int getRemaining() { return remaining; }
        public int getTotalAffected() { return totalAffected; }
    }

    public static int batchDelete(
            DataSource pool, String table, String condition,
            int batchSize, int maxRetries, long baseDelayMs)
            throws SQLException, OccRetry.MaxRetriesExceededException {

        if (batchSize < 1) throw new IllegalArgumentException("batchSize must be >= 1");

        int totalDeleted = 0;
        int consecutiveFailures = 0;
        while (true) {
            try (Connection conn = pool.getConnection()) {
                conn.setAutoCommit(false);
                String sql = "DELETE FROM " + table + " WHERE id IN (SELECT id FROM " + table
                        + " WHERE " + condition + " LIMIT " + batchSize + ")";

                // Commit is inside the retry scope so commit-time 40001 is retried.
                int deleted = OccRetry.executeWithRetry(conn, (c) -> {
                    try (Statement stmt = c.createStatement()) {
                        int count = stmt.executeUpdate(sql);
                        c.commit();
                        return count;
                    }
                }, maxRetries, baseDelayMs);

                totalDeleted += deleted;
                consecutiveFailures = 0;
                logger.info("Deleted " + deleted + " rows (total: " + totalDeleted + ")");
                if (deleted == 0) break;
            } catch (OccRetry.MaxRetriesExceededException e) {
                consecutiveFailures++;
                logger.warning("Batch OCC retries exhausted (" + consecutiveFailures
                        + "/" + MAX_CONSECUTIVE_FAILURES + "), retrying batch with fresh connection");
                if (consecutiveFailures >= MAX_CONSECUTIVE_FAILURES) throw e;
            }
        }

        // Post-verification: ensure no matching rows remain
        try (Connection conn = pool.getConnection();
             Statement stmt = conn.createStatement();
             var rs = stmt.executeQuery("SELECT COUNT(*) FROM " + table + " WHERE " + condition)) {
            if (rs.next()) {
                int remaining = rs.getInt(1);
                if (remaining > 0) {
                    throw new IncompleteBatchException(remaining, totalDeleted);
                }
            }
        }

        return totalDeleted;
    }

    public static int batchDelete(DataSource pool, String table, String condition)
            throws SQLException, OccRetry.MaxRetriesExceededException {
        return batchDelete(pool, table, condition, 1000, 3, 100);
    }

    public static int parallelBatchDelete(
            DataSource pool, String table, String condition,
            int numWorkers, int batchSize, int maxRetries, long baseDelayMs)
            throws SQLException, OccRetry.MaxRetriesExceededException {

        if (numWorkers < 1) throw new IllegalArgumentException("numWorkers must be >= 1");
        if (batchSize < 1) throw new IllegalArgumentException("batchSize must be >= 1");

        ExecutorService executor = Executors.newFixedThreadPool(numWorkers);
        List<Future<Integer>> futures = new ArrayList<>();

        for (int i = 0; i < numWorkers; i++) {
            final int workerId = i;
            futures.add(executor.submit(() -> {
                int totalDeleted = 0;
                int consecutiveFailures = 0;
                // Cast hashtext to bigint before abs() to prevent overflow on INT_MIN.
                String partitionCondition = "(" + condition + ")"
                        + " AND abs(hashtext(CAST(id AS text))::bigint) % " + numWorkers + " = " + workerId;

                while (true) {
                    try (Connection conn = pool.getConnection()) {
                        conn.setAutoCommit(false);
                        String sql = "DELETE FROM " + table + " WHERE id IN (SELECT id FROM " + table
                                + " WHERE " + partitionCondition + " LIMIT " + batchSize + ")";

                        // Commit inside retry scope.
                        int deleted = OccRetry.executeWithRetry(conn, (c) -> {
                            try (Statement stmt = c.createStatement()) {
                                int count = stmt.executeUpdate(sql);
                                c.commit();
                                return count;
                            }
                        }, maxRetries, baseDelayMs);

                        totalDeleted += deleted;
                        consecutiveFailures = 0;
                        logger.info("Worker " + workerId + ": Deleted " + deleted + " rows (total: " + totalDeleted + ")");
                        if (deleted == 0) break;
                    } catch (OccRetry.MaxRetriesExceededException e) {
                        consecutiveFailures++;
                        logger.warning("Worker " + workerId + ": Batch OCC retries exhausted ("
                                + consecutiveFailures + "/" + MAX_CONSECUTIVE_FAILURES
                                + "), retrying batch with fresh connection");
                        if (consecutiveFailures >= MAX_CONSECUTIVE_FAILURES) throw e;
                    }
                }
                return totalDeleted;
            }));
        }

        executor.shutdown();
        int total = 0;
        for (Future<Integer> future : futures) {
            try {
                total += future.get();
            } catch (ExecutionException e) {
                Throwable cause = e.getCause();
                if (cause instanceof SQLException) throw (SQLException) cause;
                if (cause instanceof OccRetry.MaxRetriesExceededException) throw (OccRetry.MaxRetriesExceededException) cause;
                throw new SQLException("Worker thread failed", cause);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                throw new SQLException("Parallel delete interrupted", e);
            }
        }

        logger.info("Parallel delete complete: " + total + " rows deleted by " + numWorkers + " workers");

        // Post-verification: ensure no matching rows remain (uses original condition, not partitioned)
        try (Connection conn = pool.getConnection();
             Statement stmt = conn.createStatement();
             var rs = stmt.executeQuery("SELECT COUNT(*) FROM " + table + " WHERE " + condition)) {
            if (rs.next()) {
                int remaining = rs.getInt(1);
                if (remaining > 0) {
                    throw new IncompleteBatchException(remaining, total);
                }
            }
        }

        return total;
    }

    public static int parallelBatchDelete(DataSource pool, String table, String condition)
            throws SQLException, OccRetry.MaxRetriesExceededException {
        return parallelBatchDelete(pool, table, condition, 4, 1000, 3, 100);
    }
}
